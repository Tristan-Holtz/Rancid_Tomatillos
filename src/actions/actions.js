export const setMovies = movies => ({
  type: 'SET_MOVIES',
  movies
});

export const setUser = user => ({
  type: 'SET_USER',
  user
})

export const setRatings = ratings => ({
  type: 'SET_RATINGS',
  ratings
})
