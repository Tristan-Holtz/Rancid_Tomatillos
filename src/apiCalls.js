export const getUser = async (email, password) => {
  const url = 'https://rancid-tomatillos.herokuapp.com/api/v1/login';
  const options = {
    method: 'POST',
    body: JSON.stringify({
      email,
      password
    }),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const res = await fetch(url, options)
  if (!res.ok) {
    throw new Error("Error! No 200 Status Code Found.")
  }
  const user = await res.json();
  return user
}

export const getMovies = async () => {
  return fetch('https://rancid-tomatillos.herokuapp.com/api/v1/movies')
    .then(response => response.json())
    .then(movies => movies.movies);
};

export const addUserRating = async (userID, movie_id, rating) => {
  const url = `https://rancid-tomatillos.herokuapp.com/api/v1/users/${userID}/ratings`;
  const options = {
    method: 'POST',
    body: JSON.stringify({
      movie_id,
      rating
    }),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const res = await fetch(url, options)
  if (!res.ok) {
  throw new Error("Error! No 200 Status Code Found.")
  }
  const updatedRating = await res.json();
  return updatedRating;
}

export const getRatings = async (userID) => {
  return fetch(`https://rancid-tomatillos.herokuapp.com/api/v1/users/${userID}/ratings`)
    .then(response => response.json())
    .then(data => data)
}

export const deleteRating = async (userID, movieID) => {
  console.log('in fetch', userID, movieID)
  return fetch(`https://rancid-tomatillos.herokuapp.com/api/v1/users/${userID}/ratings/${movieID}`, {method: 'DELETE'})
    .then(res => res.json())
    .then(data => data)
    .catch(error => console.log(error))
}