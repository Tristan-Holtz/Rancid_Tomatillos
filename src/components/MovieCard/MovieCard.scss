.card-and-rating {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 1% 0%;
}

.movie-card {
  border: 3px solid black;
  height: 400px;
  width: 280px;
  margin: 15px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  position: relative;
  box-shadow: 3px 3px 5px 5px #ffc5aa, -2px -2px 5px 5px white;
  text-align: center;
  overflow: hidden;
  transition: 0.5s ease-in-out;
}

.movie-card-image {
  height: 100%;
  width: 100%;
  transition: 0.2s;
  -webkit-transition: 0.2s;
}

.movie-card-info:hover ~ .movie-card-image {
  height: 100%;
  width: 105%;
  margin-left: -2.5%;
  transition: 0.4s ease;
}

.movie-card-info {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  color: white;
  opacity: 0%;
  height: 100%;
  width: 100%;
  position: absolute;
  top: 0px;
  box-shadow: inset 3px 3px 15px 5px #000000,
              inset -2px -2px 15px 5px #000000;
  text-shadow: 2px 2px 4px black;
  transition: 1s ease-in-out;
  -webkit-transition: 1s ease-in-out;
  overflow: auto;
  >h1 {
    font-size: 1.8em;
    padding: 3%;
  }
  >div {
    >h3 {
      font-size: 2em;
      padding: 3%;
    }
  }
}

.movie-card-info:hover {
  background-color: rgba(0,0,0,0.4);
  opacity: 100%;
  transition: 0.2s ease-in-out;
  -webkit-transition: 0.2s ease-in-out;
}

.rating-dropdown {
}

.rating-label {
  display: inline;
}
