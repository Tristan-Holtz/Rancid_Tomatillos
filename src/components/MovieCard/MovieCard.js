import React from 'react';
import './MovieCard.scss';
import { Link } from 'react-router-dom';

import { connect } from 'react-redux';
import { setRatings } from '../../actions/actions';
import { addUserRating, getRatings, deleteRating } from '../../apiCalls';

export const MovieCard = ({ movie, setRatings, user, ratings }) => {
  const submitUserRating = async event => {
    let userID = user.id;
    let movieRating = parseInt(event.target.value);
    let movieID = parseInt(event.target.id);
    await addUserRating(userID, movieID, movieRating);
    getSetRatings();
  };

  const getSetRatings = async () => {
    const ratings = await getRatings(user.id);
    let userRatings = ratings.ratings;
    setRatings(userRatings);
  };

  const checkIfRated = (movie) => {
    let movieRating = ratings.find(rating => {
      return rating.movie_id === movie.id
    });
    if(movieRating) {
      return (
        <div className='complete-rating'>
          <p className='rating-label'><span>Your Rating:</span> {movieRating.rating}/10</p>
          <button
            className='change-rating-btn'
            id={movieRating.id}
            onClick={
              (e) => {removeRating(e)}
            }
          >
            (<span id={movieRating.id}>Change rating</span>)
          </button>
        </div>
      )
    } else {
      return (
        <div className='user-rating'>
          <select
            id={movie.id}
            className='rating-dropdown'
            onChange={e => {
              submitUserRating(e);
            }}
          >
            <option>Rate This Movie</option>
            <option>1</option>
            <option>2</option>
            <option>3</option>
            <option>4</option>
            <option>5</option>
            <option>6</option>
            <option>7</option>
            <option>8</option>
            <option>9</option>
            <option>10</option>
          </select>
        </div>
      );
    }
  };

  const removeRating = async event => {
    let ratingID = event.target.id;
    await deleteRating(user.id, ratingID);
    await getSetRatings();
  };

  return (
    <article className="card-and-rating">
      <article className="movie-card" key={movie.id}>
        <Link
          to={{
            pathname: `/movies/${movie.id}`,
            state: movie
          }}
        >
          <div className="movie-card-info">
            <h1>{movie.title}</h1>
            <div className="tomato-rating">
              {movie.average_rating >= 6 && <img className="tomato-logo" src={process.env.PUBLIC_URL + '/tomato-logo.png'} alt="tomato logo" />}
              <h3 className="avg-rating-card">{Math.round(movie.average_rating * 10) / 10}/10</h3>
            </div>
          </div>
          <img className="movie-card-image" src={movie.poster_path} />
        </Link>
      </article>
      {(user) && checkIfRated(movie)}
    </article>
  );
};

export const mapStateToProps = state => ({
  movies: state.movies,
  user: state.user,
  ratings: state.ratings
});

export const mapDispatchToProps = dispatch => ({
  setRatings: ratings => {
    dispatch(setRatings(ratings));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieCard);
