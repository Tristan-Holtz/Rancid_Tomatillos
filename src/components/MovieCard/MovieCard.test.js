import React from 'react';
import { shallow } from 'enzyme';
import { MovieCard, mapStateToProps, mapDispatchToProps } from './MovieCard';
import { setRatings } from '../../actions/actions';
import { addUserRating, getRatings, deleteRating } from '../../apiCalls';

describe('MovieCard', () => {
  let wrapper;
  let mockProps;
  beforeEach(() => {
    mockProps = {
      user: { id: 1 },
      movie: { id: 2, average_rating: 5, poster_path: '' },
      ratings: [{ movie_id: 1 }, { movie_id: 2 }],
      setRatings: jest.fn(),
      event: {
        target: {
          value: 6,
          id: 2
        }
      }
    };
    wrapper = shallow(<MovieCard {...mockProps} />);
  });

  it('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('mapStateToProps', () => {
    it('should return an object with movies, user and ratings', () => {
      const mockState = {
        movies: ['movie', 'movie', 'movie'],
        user: { name: 'Greg', password: 'abc123', email: 'greg@turing.io' },
        ratings: [1, 2, 3, 4, 5],
        banana: 'bananas',
        potatoes: 'potatoes'
      };
      const expected = {
        movies: ['movie', 'movie', 'movie'],
        user: { name: 'Greg', password: 'abc123', email: 'greg@turing.io' },
        ratings: [1, 2, 3, 4, 5]
      };
      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch with the setRatings action when getSetRatings is called', () => {
      const mockDispatch = jest.fn();
      const ratings = { ratings: [1, 2, 3, 4, 5] };
      const actionToDispatch = setRatings(ratings);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.setRatings(ratings);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});
