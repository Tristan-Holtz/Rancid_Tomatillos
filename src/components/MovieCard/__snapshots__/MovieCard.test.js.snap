// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard should render the component 1`] = `ShallowWrapper {}`;
